docs/source/adrs/0007-use-concourse-to-run-e2e-tests.md

```diff
@@ -0,0 +1,52 @@
+---
+hide_in_navigation: true
+---
+
+# ADR 7: Use Concourse to run E2E tests
+
+Date: 2020-03-10
+
+## Status
+
+Accepted
+
+## Context
+
+In moving Pay's CI pipeline away from Jenkins, we need to find a new home for our endtoend tests.
+Pays end to end tests are currently run on a very big Jenkins box, using docker compose to orchestrate the platform under test.
+This means they require a large amount of compute - the box they currently run on
+is huge, and we are still limited to only being able to run 3 e2e jobs at a time.
+
+As part of the migration, Pay has built a build and deploy pipeline on the GDS shared Concourse. 
+We also attempted to put the simpler parts of our testing piepline on Concourse
+but found that to be quite complicated, so switched to using Travis to run our unit tests.
+This means at the moment the whole pipeline is split across Travis and Concourse.
+
+Pay obviously needs tests to run in a timely and reliable way. Currently on 
+Jenkins the longest running e2e test suite takes around 4 minutes
+and reliability is around 90%. Any replacement should improve on these figures.
+
+
+We carried out some spikes into different ways we could run the e2e tests:
+1. Using Travis <https://payments-platform.atlassian.net/browse/PP-6042>
+1. Using Concourse to run the tests against a target deployed on the PaaS <https://payments-platform.atlassian.net/browse/PP-6206>
+1. Using Concourse to run the tests against a target deployed on docker-compose
+1. Using Buildkite <https://payments-platform.atlassian.net/browse/PP-6214>
+
+For detailed discussion of the outcomes of the spikes see the relevant tickets.
+
+The conclusion from these spikes was that options 3 and 4 were both viable. 
+Option 3 is the most inline with GDS tech direction, but requires a bit more work to get the required performance - it is
+possible to run many builds in parralel on separate worker boxes, but this comes less natively than in option 4.
+Otion 4 is a good technological fit, but requires more work to get everything set up well, and is a distinct divergence from
+current GDS technology.
+
+## Decision
+We will run the e2e tests in docker compose (as they currently are), on Concourse (option 3 above).
+We will also move the rest of Pay's tests on to Concourse.
+We will arrange for a sufficient number of concourse workers to be proviioned for Pay to ensure we are not blocked by e2e tests.
```

Comment:
do we know how concourse handles propagating intermediate build results between nodes? or is it that a pipeline runs on a single worker node and data isn't move around?

Do we know roughly how we'll structure the pipelines in terms of which types of tests would be run, and how? eg. unit, pact, e2e? It's totally fine if we don't know the answers to these questions now but it would be helpful to note what we do know and what we don't know yet.

Reply:
I don't know exactly how we'll structure the pipelines at this point. But E2E tests should be able to fit in to the pipelines we have already built without too much issue.

Reply:
Concourse is able to produce artefacts after a build stage. However the instance on which a build runs is ephemeral. For this reason Tech Ops have provided a "local" ECS repo to which candidate Docker images can be pushed for reuse elsewhere in a CI pipeline without needing to use Docker hub. Authentication is handled automatically. This provides similar functionality to locally tagging builds on a single Jenkins box, which wouldn't scale anyway if we used Jenkins worker nodes.